cpp/include/raft/neighbors/detail/ivf_pq_serialize.cuh

```diff
@@ -0,0 +1,221 @@
+
+template <typename IdxT, typename SizeT>
+void serialize_list(const raft::device_resources& handle,
+                    std::ostream& os,
+                    const list_data<IdxT, SizeT>& ld,
+                    std::optional<SizeT> size_override = std::nullopt)
+{
+  auto size = size_override.value_or(ld.size.load());
+  serialize_scalar(handle, os, size);
+  if (size == 0) { return; }
+
+  auto data_extents = make_extents<SizeT>(div_rounding_up_safe<SizeT>(size, kIndexGroupSize),
+                                          ld.data.extent(1),
+                                          ld.data.extent(2),
+                                          ld.data.extent(3));
+  auto data_array   = make_host_mdarray<uint8_t, SizeT, row_major>(data_extents);
+  auto inds_array   = make_host_mdarray<IdxT, SizeT, row_major>(make_extents<SizeT>(size));
+  copy(data_array.data_handle(), ld.data.data_handle(), data_array.size(), handle.get_stream());
+  copy(inds_array.data_handle(), ld.indices.data_handle(), inds_array.size(), handle.get_stream());
+  handle.sync_stream();
+  serialize_mdspan(handle, os, data_array.view());
+  serialize_mdspan(handle, os, inds_array.view());
+}
+
+template <typename IdxT, typename SizeT>
+void serialize_list(const raft::device_resources& handle,
+                    std::ostream& os,
+                    const std::shared_ptr<const list_data<IdxT, SizeT>>& ld,
+                    std::optional<SizeT> size_override = std::nullopt)
+{
+  if (ld) {
+    return serialize_list(handle, os, *ld, size_override);
+  } else {
+    return serialize_scalar(handle, os, SizeT{0});
+  }
+}
+
+template <typename IdxT, typename SizeT>
+void deserialize_list(const raft::device_resources& handle,
+                      std::istream& is,
+                      std::shared_ptr<list_data<IdxT, SizeT>>& ld,
+                      uint32_t pq_bits,
+                      uint32_t pq_dim)
+{
+  auto size = deserialize_scalar<SizeT>(handle, is);
+  if (size == 0) { return ld.reset(); }
+  std::make_shared<list_data<IdxT, SizeT>>(handle, size, pq_bits, pq_dim).swap(ld);
+  auto data_extents = make_extents<SizeT>(div_rounding_up_safe<SizeT>(size, kIndexGroupSize),
+                                          ld->data.extent(1),
+                                          ld->data.extent(2),
+                                          ld->data.extent(3));
+  auto data_array   = make_host_mdarray<uint8_t, SizeT, row_major>(data_extents);
+  auto inds_array   = make_host_mdarray<IdxT, SizeT, row_major>(make_extents<SizeT>(size));
+  deserialize_mdspan(handle, is, data_array.view());
+  deserialize_mdspan(handle, is, inds_array.view());
+  copy(ld->data.data_handle(), data_array.data_handle(), data_array.size(), handle.get_stream());
+  // NB: copying exactly 'size' indices to leave the rest 'kInvalidRecord' intact.
+  copy(ld->indices.data_handle(), inds_array.data_handle(), size, handle.get_stream());
+}
```

Comment:
>[cudaMemcpyAsync()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79) is asynchronous with respect to the host, so the call may return before the copy is complete.

You are executing a H2D copy and the host array could go out of scope before the copy is complete. Yes, I think this means we should sync here.

Reply:
The rmm::device_buffers, which are in the back of our `device_mdarray`, seem to deallocate the memory [asynchronously](https://github.com/rapidsai/rmm/blob/c9c83abb3c5b8285de98aad9c0ebe4f988015029/include/rmm/device_buffer.hpp#L232-L237), so that shouldn't be necessary (otherwise we really need to force sync in many places in raft!).

Reply:
Ah, I get in now, you mean the problem is that these two are the host arrays not aware of cuda streams? Good point, I'll add the synchronize and check if I have this pattern anywhere else in the PR.